pkg/serviceaccount/claims.go

```diff
@@ -36,25 +41,27 @@ type privateClaims struct {
 }
 
 type kubernetes struct {
-	Namespace string `json:"namespace,omitempty"`
-	Svcacct   ref    `json:"serviceaccount,omitempty"`
-	Pod       *ref   `json:"pod,omitempty"`
-	Secret    *ref   `json:"secret,omitempty"`
+	Namespace string          `json:"namespace,omitempty"`
+	Svcacct   ref             `json:"serviceaccount,omitempty"`
+	Pod       *ref            `json:"pod,omitempty"`
+	Secret    *ref            `json:"secret,omitempty"`
+	WarnAfter jwt.NumericDate `json:"warnafter,omitempty"`
 }
 
 type ref struct {
 	Name string `json:"name,omitempty"`
 	UID  string `json:"uid,omitempty"`
 }
 
-func Claims(sa core.ServiceAccount, pod *core.Pod, secret *core.Secret, expirationSeconds int64, audience []string) (*jwt.Claims, interface{}) {
+func Claims(sa core.ServiceAccount, pod *core.Pod, secret *core.Secret, expirationSeconds int64, audience []string, extendExpiry bool) (*jwt.Claims, interface{}, jwt.NumericDate) {
```

Comment:
I'd actually move the logic around the expiry and warn deadline to the caller of this function, rather than doing complex logic inside this function and plumbing conditional expirations back out in return values.

Proposed signature:
`func Claims(sa core.ServiceAccount, pod *core.Pod, secret *core.Secret, expirationSeconds, warnSeconds int64, audience []string) (*jwt.Claims, interface{})`

Reply:
Then this function would just do

```
if warnSeconds != 0 {
  pc.Kubernetes.WarnAfter = jwt.NewNumericDate(now.Add(time.Duration(warnSeconds) * time.Second))
}
```

Reply:
Well actually I wanted to modify the fields in the caller initially, but since the private claims are returned as a general interface{} object, I need to do something like
```
private, ok := privateObj.(*privateClaims)
	if !ok {
		klog.Errorf("jwt validator expected private claim of type *privateClaims but got: %T", privateObj)
		return nil, errors.New("Token could not be validated.")
	}
```
to parse the claim struct, push in the warnafter field, while also modifying the public claim's expiry field. This makes me tend to push all the logic into the Claims method

Reply:
I'm not sure I understand. Keeping the return values opaque, and taking an additional warn-after parameter, and lengthening the expiration time at the call site keeps the extension logic in one place and avoids computing a value in one place and then overriding based on the returned result.

Reply:
Ah now I get what you mean, I was thinking in a different direction. Refactored.
(I still put these constants in serviceaccount package, since it is a easier place to import them compared to the calling site)